packages/core/saved-objects/core-saved-objects-server/src/saved_objects_type.ts

```diff
@@ -128,6 +132,102 @@ export interface SavedObjectsType<Attributes = any> {
+   * by exposing an unified way of describing the changes of shape or data of the type.
+   *
+   * Model versioning is decoupled from Kibana versioning, and isolated between types.
+   * Model versions are identified by a single numeric value, starting at `1` and without gaps.
+   *
+   * Please refer to {@link SavedObjectsModelVersion} for details on the API's usages.
+   *
+   * A **valid** versioning would be:
+   *
+   * ```ts
+   * {
+   *   name: 'foo',
+   *   // other mandatory attributes...
+   *   modelVersions: {
+   *     '1': modelVersion1,
+   *     '2': modelVersion2,
+   *     '3': modelVersion3,
+   *   }
+   * }
+   * ```
+   *
+   * A **invalid** versioning would be:
+   *
+   * ```ts
+   * {
+   *   name: 'foo',
+   *   // other mandatory attributes...
+   *   modelVersions: {
+   *     '1': modelVersion1,
+   *     '3': modelVersion3, // ERROR, no model version 2
+   *     '3.1': modelVersion31, // ERROR, model version is a single numeric value
+   *   }
+   * }
+   * ```
+   *
+   * @alpha experimental and subject to change.
+   */
+  modelVersions?: SavedObjectsModelVersionMap | SavedObjectsModelVersionMapProvider;
+
+  /**
+   * Allows to opt-in to the new model version API.
+   *
+   * Must be a valid semver version (with the patch version being necessarily 0)
+   *
+   * When specified, the type will switch from using the {@link SavedObjectsType.migrations | legacy migration API}
+   * to use the {@link SavedObjectsType.modelVersions | modelVersion API} after the specified version.
+   *
+   * When opted in, it will no longer be possible to use the legacy migration API after the specified version.
+   *
+   * A **valid** usage example would be:
+   *
+   * ```ts
+   * {
+   *   name: 'foo',
+   *   // other mandatory attributes...
+   *   switchToModelVersionAt: '8.8.0',
+   *   migrations: {
+   *     '8.1.0': migrateTo810,
+   *     '8.7.0': migrateTo870,
+   *   },
+   *   modelVersions: {
+   *     '1': modelVersion1
+   *   }
+   * }
+   * ```
+   *
+   * An **invalid** usage example would be:
+   *
+   * ```ts
+   * {
+   *   name: 'foo',
+   *   // other mandatory attributes...
+   *   switchToModelVersionAt: '8.9.0',
+   *   migrations: {
+   *     '8.1.0': migrateTo8_1,
+   *     '8.9.0': migrateTo8_9, // error: migration registered for the switch version
+   *     '8.10.0': migrateTo8_10, // error: migration registered for after the switch version
+   *   },
+   *   modelVersions: {
+   *     '1': modelVersion1
+   *   }
+   * }
+   * ```
+   *
+   * Please refer to the {@link SavedObjectsType.modelVersions | modelVersion API} for more documentation on
+   * the new API.
+   *
+   * @remarks All types will be forced to switch to use the new API in a later version. This switch is
+   *          allowing types owners to switch their types before the milestone.
+   */
+  switchToModelVersionAfter?: string;
```

Comment:
Since `swithToModelVersionAfter` actually means changing to the new API, how about using `switchToModelVersionMigrationStrategyAfter` rather? It's a much longer name but does clearly distinguish what the property refers to.